Fix bugs

modos189 opened the pull request:
Library is not compiled because of a typo. Fixed, plus added extra check to prevent errors.

modos189 committed:
Fix typo

johnd0e commented:
@Spaction 
https://github.com/Spaction/Leaflet.Canvas-Markers/commit/153525e045e1c4ec22a0ecb02d965366e99852a0 introduced regression in event handling, open Issues section to discuss.

Commit:
fix broken removeMarker

johnd0e commented:
> Additional checks to prevent errors during atypical use

`_searchPoints` is called from mouse event listeners, and we have exception if `CanvalLayerMarker` is empty.
It could be fixed in some another place, e.g. do not attach listeners until we have markers, or check markers existence in the beginning of listener.

Commit:
fix exception in mouse event handler with empty CanvasMarkerLayer

`this._markers` is undefined in `_searchPoints`, called from event listeners.
Now we check markers existence in the beginning of handlers.

Note 1:
Could be fixed in some another place:
 e.g. check `_markers` in `_searchPoints`
 ...or initialize (empty) `_markers` on CanvasMarkerLayer init

Note 2:
`!this._map` removed from condition in `_onMouseMove:` 'cause
it's never should be true as we detach listeners on remove from map

johnd0e commented:
Also see https://github.com/IITC-CE/Leaflet.Canvas-Markers/issues/10

johnd0e commented:
https://github.com/Spaction/Leaflet.Canvas-Markers/blob/153525e045e1c4ec22a0ecb02d965366e99852a0/src/plugin/leaflet.canvas-markers.js#L135-L139
`size` is already defined.
`size` and `m` assigned a value but never used.

---
https://github.com/Spaction/Leaflet.Canvas-Markers/blob/153525e045e1c4ec22a0ecb02d965366e99852a0/src/plugin/leaflet.canvas-markers.js#L562
`iconOptions` is assigned a value but never used.

---
https://github.com/Spaction/Leaflet.Canvas-Markers/blob/153525e045e1c4ec22a0ecb02d965366e99852a0/src/plugin/leaflet.canvas-markers.js#L281
`self` is not needed here as it is only used in the same scope (with the same `this`).

johnd0e commented:
https://github.com/Spaction/Leaflet.Canvas-Markers/blob/153525e045e1c4ec22a0ecb02d965366e99852a0/src/plugin/leaflet.canvas-markers.js#L494-L501

`redraw` is not defined.

Commit:
removeMarker: fix exception when CanvasMarkerLayer is not attached to map

johnd0e commented:
Concerning last fix (ab64e18):  in several places of code there are excessively verbose variant:
```
            if (self._map)
                isDisplaying = self._map.getBounds().contains(latlng);
            else
                isDisplaying = false;
```

Similar excessive patterns present in many places, e.g. 
```
            if (redraw && redraw === true)
```
```
            if (isDisplaying === true && redraw === true) {
```

Commit:
fix exception if image is not loaded (yet) / bad url / etc

Was protected in marker add time, but not on zoom/pan

Sample:
```
VM2992:16252 Uncaught DOMException: Failed to execute 'drawImage' on 'CanvasRenderingContext2D': The HTMLImageElement provided is in the 'broken' state.
    at NewClass._drawImage (<anonymous>:16252:23)
    at NewClass._drawMarker (<anonymous>:16239:22)
    at <anonymous>:16200:22
    at Array.forEach (<anonymous>)
    at NewClass._draw (<anonymous>:16179:16)
    at NewClass._redraw (<anonymous>:16077:18)
    at NewClass.fire (<anonymous>:1867:11)
    at NewClass._move (<anonymous>:5506:9)
    at NewClass._onZoomTransitionEnd (<anonymous>:5963:8)
```

johnd0e commented:
> fix exception if image is not loaded (yet) / bad url / etc

I suppose there is one more fix needed: invalidate delayed draw queue (`self._imageLookup[iconUrl][2]`) on pan/zoom.
Small refactoring is needed to do it more effectively (separate queue from img cache).
